tests/README.md

```diff
@@ -182,27 +141,45 @@ expected output. The framework includes the following built-in verify functions:
    1. For each line in `expected` with the prefix "- " there must be no line in
       the output containing the following string.
 
-1. `_verify_failure`** `func`
+1. **`_verify_failure`** `func`
 
    Runs `func` and confirms that it fails (i.e., non-zero return code). This function is useful
    for testing commands that demonstrate configurations that are expected to fail.
 
-## Running the Tests: Make
+After all test steps are run, add
+```sh
+# @cleanup
+```
+as a single line followed by the cleanup steps that clean up all the resources. These steps can also directly call functions defined in the auto-sourced scripts as described before. But others like `util/samples.sh` need to be sourced again in cleanup, even if you have already sourced them for the test steps.
```

Comment:
> But others like `util/samples.sh` need to be sourced again in cleanup

That seems kind of ugly. Won't the bash linter complain about that? How about the framework look for any `source` lines in the main script, and put a copy in the cleanup script?

Reply:
Another sort of related thought ... would it make sense to leave the cleanup steps in the main bash script, in addition to putting a copy in the cleanup script? Then in the "successful completion" (common) case, the test is completely finished after the main script completes. Only in the failure case, would you have to run the cleanup script. That would seem nicer to me (and more efficient, since you won't need to re-source all the includes).

Reply:
Personally I'd prefer the first approach. The cleanup steps will be run anyway as they are `defer`-ed, and re-sourcing files are only minimal overhead compared to test themselves.

Reply:
If you leave the cleanup steps in the main script, you would not `defer` the cleanup script. Instead you would just run the main script, if success you're done, if failure you run the cleanup script. I'm not sure, though, if the underlying istio test framework gives you the hooks to do that, but if it does, that does seem nicer to me.

That said, it's not a big deal, so whatever you decide is fine.

Reply:
This is now resolved with the first solution as the istio test framework doesn't seem to allow that much flexibility.